0010-B-object-storage-building-block.md

```diff
@@ -0,0 +1,275 @@
+
+A successful response will be:
+
+- Status: `200 OK`
+- Headers: each metadata key is a header; see below for a comment on headers. Additionally, the `ETag` is passed as response header too.
+- Body: the body of the response is the raw bytes of the object, which are sent to the client in a streamed way, as soon as they are retrieved from the state store (that is: Dapr does not buffer the data in-memory before sending it to the client).
+
+An error response will have:
+
+- Status:
+  - `404 Not Found` for objects not found
+  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
+  - `500 Internal Server Error` for all other errors
+- Headers: response will have `Content-Type: application/json`
+- Body: a JSON-encoded payload including more details on the error
+
+#### Storing an object
+
+To store an object, clients make a PUT request:
+
+```text
+PUT http://localhost:<port>/v1.0-alpha1/objects/<component>/<key>
+```
+
+Where `<component>` is the name of the object store component, and `<key>` is the key (name) of the object.
+
+The request body contains the raw bytes of the object, which are read by Dapr in a streamed way and are sent to the object store as soon as they are available (once again, Dapr does not buffer the data in-memory before storing it).
+
+Additionally, metadata keys are passed as header values; see below for a comment on headers. An `ETag` can be passed as request header too for concurrency control, and it's ignored if the object doesn't exist already. If an ETag is not specified, the write will always succeed (unless other errors occur).
+
+A response will be:
+
+- Status:
+  - `200 OK` for successful operations
+  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
+  - `409 Conflict` if the object already exists and there's an ETag mismatch
+  - `500 Internal Server Error` for all other errors
+- Headers: response will have `Content-Type: application/json`
+- Body:
+  - For successful responses: a JSON-encoded payload containing the ETag and the number of bytes stored.
+  - In case of errors: a JSON-encoded payload including more details on the error
+
+#### Deleting an object
+
+To delete an object, clients make a DELETE request:
+
+```text
+DELETE http://localhost:<port>/v1.0-alpha1/objects/<component>/<key>
+```
+
+Where `<component>` is the name of the object store component, and `<key>` is the key (name) of the object.
+
+An `ETag` can be passed as request header too for concurrency control. If an ETag is is specified, the operation will fail if the ETag doesn't match; the ETag is optional and when it's missing there's no concurrency control.
+
+A response will be:
+
+- Status:
+  - `204 No Content` for successful operations
+  - `400 Bad Request` if the state store doesn't exist or the key parameter is missing
+  - `409 Conflict` if the object already exists and there's an ETag mismatch
+  - `500 Internal Server Error` for all other errors
```

Comment:
What about write-only object storage?

Reply:
What do you mean?

Reply:
For example, an HSM might be backing this that allows storing new key material, and reading public keys, but not deletions. Maybe the object is backed by a file system but the file is marked read-only.

I'd suggest a 403: forbidden, to handle the case where deletion isn't allowed.

I'd also recommend a 410 if the content is already deleted and no longer exists.

Reply:
Support for HSM is out of scope of this building block. It would more likely be aligned with the crypto and/or secret stores building blocks.

But in general, i support adding more status codes to indicate various scenarios like the ones you describe.

Reply:
> Support for HSM is out of scope of this building block.

I was illustrating the error, but yeah, I agree.